Update README.md (#39)
dangowrt/linksys-e8450-openwrt-installer: README.md

**WARNING** This will replace the bootloader (TF-A 2.4, U-Boot 2021.10) and convert the flash layout of the device to UBI! The installer stores a copy of the previous bootchain in a dedicated UBI volume `boot_backup`. If you want to go back to the vendor firmware, you will have to boot into recovery mode (ie. initramfs),
copy those files into `/tmp`, umount the UBI volume, detach the UBI device and then write the files to the corresponding MTD partitions (mtd write $file /dev/mtdX).
The resulting file `openwrt-mediatek-mt7622-linksys_e8450-ubi-initramfs-recovery-installer.itb` can be flashed via the vendor/official firmware web interface, as well as OpenWrt firmware running non-UBI builds (by running `sysupgrade -F openwrt-mediatek-mt7622-linksys_e8450-ubi-initramfs-recovery-installer.itb`).

To be on the safe side, it is still recommended to make a complete backup of the device flash __**before**__ running the installer!
**WARNING #1** This will replace the bootloader (TF-A 2.4, U-Boot 2021.10) and convert the flash layout of the device to UBI! The installer stores a copy of the previous bootchain in a dedicated UBI volume `boot_backup`.

> To use the script to generate the installer image, you will need all runtime requirements of the OpenWrt ImageBuilder installed as well as `libfdt-dev` and `cmake` which is needed by the installer-generator itself.

Reviewer: You are removing the requirement for the **run-time dependencies** of the ImageBuilder here...

> * OpenWrt ImageBuilder

Reviewer: and then you are adding the ImageBuilder itself as a requirement here. Better
```suggestion
 * all [prerequisites of the OpenWrt ImageBuilder](https://openwrt.org/docs/guide-user/additional-software/imagebuilder#prerequisites) 
```

## Installing OpenWrt
#### Assuming the device is running stock firmware, and is brand new or just after factory reset.

1. Connect any of the LAN ports of the device directly to the Ethernet port of your computer.
2. Set the IP address of your computer as `192.168.1.254` with netmask `255.255.255.0`, no gateway, no DNS.
3. Power on the device, wait about a minute for it to be ready.
4. Open a web browser, navigate to http://192.168.1.1 and wait for the wizard to come up.
5. Click *exactly* inside the radio button to confirm the terms and conditions, then abort the wizard.

> 6. You should then be greeted by the login screen, the stock password is "admin". (This step may be required for Linksys E8450 only)

Reviewer: ```suggestion
6. You should then be greeted by the login screen, the stock password is "admin".
```

I have personally verified this and yes, this is necessary on both, the Linksys E8450 and Belkin RT3200.

7. Navigate to __Administration__ -> __Firmware Upgrade__
8. Upload `openwrt-mediatek-mt7622-linksys_e8450-ubi-initramfs-recovery-installer.itb`.

> 9. Wait for a few minutes, the OpenWrt recovery image should come up.

Reviewer: ```suggestion
9. Wait for a minute, the OpenWrt recovery image should come up.
```

> 10. Login and navigate to __System__, save a copy of each of the `mtdblock`.

Reviewer: ```suggestion
```
I'm sorry but this doesn't make sense. You would make backups of the freshly installed bootchain which obviously works already. You should have done the backups before.

11. Upload `openwrt-mediatek-mt7622-linksys_e8450-ubi-squashfs-sysupgrade.itb`.
12. The device will reboot, you may proceed to setup OpenWrt.

> ## Upgrading to the Latest OpenWrt Snapshot Release.

Reviewer: ```suggestion
## Upgrading to the latest OpenWrt snapshot.
```
It was already wrong to call a "snapshot" a "release" as it is exactly *not* a release, but only a nightly development snapshot. Using capitals here makes it worse imho...

6. You should then be greeted by the login screen, the stock password is "admin". (This step may be required for Linksys E8450 only)
7. Navigate to __Administration__ -> __Firmware Upgrade__
8. Upload `openwrt-mediatek-mt7622-linksys_e8450-ubi-initramfs-recovery-installer.itb` to vendor web interface upgrade page.
9. Wait for OpenWrt recovery image to come up.
10. Login and navigate to __System__ -> __Backup / Flash Firmware__
11. Upload the `openwrt-mediatek-mt7622-linksys_e8450-ubi-sysupgrade.itb` file.
12. Reboot and proceed to a normal OpenWrt setup (or upload your configuration file).
8. Upload `openwrt-mediatek-mt7622-linksys_e8450-ubi-initramfs-recovery-installer.itb`.
9. Wait for a few minutes, the OpenWrt recovery image should come up.
10. Login and navigate to __System__, save a copy of each of the `mtdblock`.
11. Upload `openwrt-mediatek-mt7622-linksys_e8450-ubi-squashfs-sysupgrade.itb`.
12. The device will reboot, you may proceed to setup OpenWrt.

## Upgrading to OpenWrt snapshot release.
## Upgrading to the Latest OpenWrt Snapshot Release.

**WARNING**

SNAPSHOT RELEASES ARE LARGELY UNTESTED!

PROCEED AT YOUR OWN RISK!

> 1. If you haven't already, backup every "mtdblock" category, and move the "boot_backup" to another device. (In the event of emergency you can reflash via [JTAG](https://openwrt.org/toh/linksys/e8450#jtag)

Reviewer: I think we should still mention that users should rescue the content of `boot_backup` to another device.

> Attended Sysupgrade (auc) is included since version 0.6, all you need to do is to connect the router to the Internet and run `auc` from the web interface.

Reviewer: There are two different clients to the [asu](https://github.com/aparcar/asu):
 * `luci-app-attendedsysupgrade`: written by aparcar and providing automated upgrades using web-UI
 * `auc`: written by dangowrt and providing automated upgrades on command line interface, e.g. when connecting via SSH

Hence your wording is a bit misleading here.

**WARNING #1** This will replace the bootloader (TF-A 2.4, U-Boot 2021.10) and convert the flash layout of the device to [UBI](https://github.com/dangowrt/linksys-e8450-openwrt-installer/issues/9). The installer stores a copy of the previous bootchain in a dedicated UBI volume `boot_backup`.

**WARNING #2** Re-flashing the installer when the device is already using UBI flash layout will erase the previously backed up bootchain, which in most cases would be the vendor/official one.

For utmost safety (but not absolutely necessary), you are recommended to make a complete backup of the device flash __**before**__ running the installer. (see below "Device Flash Backup Procedure")
For utmost safety (but not absolutely necessary), you are recommended to make a complete backup of the device flash __**before**__ running the installer. (see below "Device flash backup procedure")

You'll need the below to use the script to generate the installer image:
* OpenWrt ImageBuilder
* `libfdt-dev`

> * `cmake`

Reviewer: Do not delete the requirement for `libfdt-dev` and `cmake`.
They are both required **in addition** to the prerequisites of the OpenWrt ImageBuilder.
The OpenWrt ImageBuilder itself is **not** required, it will rather be downloaded and PGP-verified by running the script.

3. Power on the device, wait about a minute for it to be ready.
4. Open a web browser, navigate to http://192.168.1.1 and wait for the wizard to come up.
5. Click *exactly* inside the radio button to confirm the terms and conditions, then abort the wizard.
6. You should then be greeted by the login screen, the stock password is "admin".
7. Navigate to __Administration__ -> __Firmware Upgrade__.
8. Upload `openwrt-mediatek-mt7622-linksys_e8450-ubi-initramfs-recovery-installer.itb`.
9. Wait for a minute, the OpenWrt recovery image should come up.
9. Navigate to __System__ -> __Backup / Flash Firmware__.
10. Upload `openwrt-mediatek-mt7622-linksys_e8450-ubi-squashfs-sysupgrade.itb`.
12. The device will reboot, you may proceed to setup OpenWrt.

## Upgrading to the latest OpenWrt snapshot 

**WARNING**
SNAPSHOT RELEASES ARE LARGELY UNTESTED!
PROCEED AT YOUR OWN RISK!

1. Backup the original/vendor bootchain

   Connect to the device via SSH and enter the following commands:

   ```
   mkdir /tmp/boot_backup

>    mount -t ubifs /dev/ubi0_3 /tmp/boot_backup

Reviewer: ```suggestion
   mount -t ubifs ubi0:boot_backup /tmp/boot_backup
```

   ```

   Then, copy the files under */tmp/boot_backup* using *scp* to your host. These files are needed in case you want to restore the original/vendor firmware. They can also be used in emergency case for reflashing via [JTAG](https://openwrt.org/toh/linksys/e8450#jtag)

> 2. `auc` (Attended Sysupgrade) and `luci-app-attendedsysupgrade` are included since version 0.6. Simply navigate to __System__ -> __Attended Sysupgrade__ and proceed accordingly.

Reviewer: ```suggestion
2. `auc` and `luci-app-attendedsysupgrade` are included since version 0.6. 
  * To update using the LuCI web interface, navigate to __System__ -> __Attended Sysupgrade__ and proceed accordingly.
  * To update using the command line use `auc`, e.g. when connecting to the router via SSH.
```

`auc` is the "**a**ttended sys**U**pgrade **c**ommand-line client".
`luci-app-attendedsysupgrade` is a LuCI web-interface based client using the same web service.
(`/etc/config/attendedsysupgrade` is the configuration)

Reviewer: :+1:

## To enter recovery mode under OpenWrt

1. Hold down the "reset" button (below the "WPS" button) whilst powering on the device.
2. Release the button once the power LED turns into orange/yellow.

This will remove any user configuration and allow restoring or upgrading from [ssh](https://openwrt.org/docs/guide-user/installation/sysupgrade.cli)/http/[tftp](https://openwrt.org/docs/guide-user/installation/generic.flashing.tftp).

## Device flash backup procedure

> 1. Flash `openwrt-mediatek-mt7622-linksys_e8450-ubi-initramfs-recovery.itb` (note that this file doesn't have the word _installer_ in it's filename)

Reviewer: ... while still running the stock firmware or non-UBI OpenWrt.

done
```

Then, copy the resulting files using *scp* to your host.

## Restoring the vendor/official firmware ##

> 1. Flash `openwrt-mediatek-mt7622-linksys_e8450-ubi-initramfs-recovery.itb` (note that this file doesn't have the word _installer_ in it's filename)

Reviewer: Since [v0.6.1](https://github.com/dangowrt/linksys-e8450-openwrt-installer/releases/tag/v0.6.1) it's enough to just enter recovery regularly, no need to flash anything at this point.
